design/dedicated-osd-pod.md

```diff
@@ -0,0 +1,147 @@
+| 5  | Detect device provisioning Configmap change, parse Configmap, extract OSD info, construct OSD Pod command and arg | | | |
+| 6  | Create one deployment per OSD | | |
+| 7  | | | | Start `ceph-osd` Daemon one Pod per device |
+
+This change addresses the above limitations in the follow ways:
+- High reliability. Each `ceph-osd` daemon runs its own Pod, their restart and upgrade are by Kubernetes controllers. Upgrading Device Provisioner Pod no longer restarts `ceph-osd` daemons.
+- More efficient resource requests. Once Device Discovery detects all devices, Rook Operator is informed of the topology and assigns appropriate resources to each Ceph OSD deployment.
+- Reusable. Device discovery can be used for other storage backends.
+
+
+### Detailed Device Discovery Process
+
+Each `Device Discovery` Pod reads device filters from `Device Filter Configmap`, retrieves named device filters, walks through device trees and uses device filters to identify the devices, and stores the filtering results in `Raw Device Configmap`. 
+
+#### Device Filter Configmap
+
+This Configmap defines the node regex and device filters on each node. 
+
+A sample of filter Configmap is as the following:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  namespace: rook-system
+  name: device-filters
+data:
+  config: |
+    - nodes: "foo*" # node name regex
+      filters: 
+      - name: "search sd*"
+        filter: by-device-path
+        parameters: 
+          - "/dev/sd*"
+    - nodes: "bar*" # node name regex
+      filters: 
+      - name: "search big disks"
+        filter: by-device-capacity
+        parameters:
+           - "gt" # greater than
+           - "100GiB" # 100GiB 
+      - name: "search vendor X"
+        filter: by-device-vendor
+        parameters:
+          - "X" # vendor name
+```
+
+In this configuration, device filters are applied to node names with match regex. Each named device filters, `by-device-path`, `by-device-capacity`, and `by-device-vendor`, are implemented by the Device Discovery Pod to conduct certain search criteria. For instance, filter `by-device-capacity` is implemented to probe each block device's and filters based on capacity given in the parameter, which in this example is `100GiB`. 
+
+Each of the named filter yields a formatted output. In the `by-device-capacity` filter, the output is a json string like the following. Note `device-path` is mandatory for all filter output.
+
+```json
+{
+ "output": [ 
+   {"device-path": "/dev/sdx", "capacity": 214748364800},
+   {"device-path": "/dev/sdy", "capacity": 214748364800}
+ ]
+}
+```
+
+#### Raw Device Configmap
+
+After devices are discovered, they are stored in `Raw Device Configmap`. A sample is as the following:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  namespace: rook-system
+  name: raw-devices
+data:
+  devices: |
+    - nodes: "foo-node1"
+      - name: "search big disks"
```

Comment:
with the filter in the cluster crd, the raw device cm wouldn't need the name and filter here. It would only need the node name and its raw devices. We would also keep track of which cluster has claimed a device so clusters don't collide with each other.

Reply:
Having filter name here helps device removal. When a cluster CRD is removed, we need to remove the devices that are used by the cluster's filter but not by others. Since Operator doesn't know, based on device path, which are the devices claimed by the filter, we need to annotate the devices so Operator can find the devices and release them if they are not referenced by other filters.

Reply:
agreed, we should use whatever annotation is needed for the full lifecycle. With the clarification on the filter above, maybe this won't be needed